rat/__init__.py

```diff
@@ -1,3 +1,145 @@
+r"""
+If you want to see a list of supported math functions/distributions, skip to [here](#language-function-reference)
+
+---
+
+## Subscripts
+
+Subscripts are the core feature that enables concise expressions. They can be seen as extensions of groups/factors in `lme4` syntax.
+
+Say we have the following data dataframe:
+```
+    game_id  home_score  away_score home_team away_team  score_diff  year
+0         1         117          88       CLE       NYK        29.0  2016
+1         2         113         104       POR       UTA         9.0  2016
+2         3         100         129       GSW       SAS       -29.0  2016
+3         4          96         108       ORL       MIA       -12.0  2016
+4         5         130         121       IND       DAL         9.0  2016
+5         6         122         117       BOS       BKN         5.0  2016
+6         7         109          91       TOR       DET        18.0  2016
+7         8          96         107       MIL       CHA       -11.0  2016
+8         9         102          98       MEM       MIN         4.0  2016
+9        10         102         107       NOP       DEN        -5.0  2016
+10       11          97         103       PHI       CLE        -6.0  2016
+```
+and we're trying to estimate some hidden skill parameter for each team. We can write the following model:
+```
+score_diff ~ normal(skills[home_team, year] - skills[away_team, year], sigma);
```

Comment:
I've been thinking we should require subscripts on data as well. So here instead of `score_diff` we have `score_diff[home_team, away_team, year]`. Just for an abundance of clarity.

Reply:
Hmm. Is it just for defining which subscripts we're going to reference on the rhs?

Reply:
Yeah something like that was the idea. Thinking more the problem is that there'd be a lot of wrong ways to index this. And I don't even think we have the syntax to make it clear what is going on. Imagine if we had multiple measurements for a single home_team, away_team, year tuple (which I guess wouldn't happen in basketball but maybe something else)

New stance is if the variable is in the dataframe, there are no subscripts.

Reply:
>  Imagine if we had multiple measurements for a single home_team, away_team, year tuple

I don't understand how this can be a problem. Wouldn't we just increment the log probability of said variable multiple times?

Reply:
Everywhere we use subscripts so far they're refer to unique elements of data structure.

By allowing subscripts on data, using the above data, we could say something like:

```
y[year] ~ ...
```

y is a column of the dataframe, so the value of y on each row is just the value of y on each row. The index is not adding anything. It makes it seem like there are only num(years) elements of y which probably isn't true.

We currently use these subscripts on the left hand side for parameters. In this case they're defining names which expressions on the right hand side can use. Maybe it makes sense to allow no subscripts on those as well in the case that the subscripts aren't used on the right hand side.

Like:
```
score_diff ~ normal(skills[year], sigma);
skills ~ normal(0, 1);
```

We might already allow this in the code -- I don't think I've thought through it tho lol.